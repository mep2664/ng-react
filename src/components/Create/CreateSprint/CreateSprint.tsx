import * as React from "react";
import { TextInput } from "../..";
import gql from "graphql-tag";
import { useMutation } from "@apollo/react-hooks";

export const CREATE_SPRINT = gql`
    mutation CREATE_SPRINT($sprintName:String!, $goal:String, $dateStart:DateTime, $dateEnd:DateTime) {
        createSprint(sprintName:$sprintName, goal:$goal, dateStart:$dateStart, dateEnd:$dateEnd){
            sprint {
                sprintId
                sprintName
                goal
                dateStart
                dateEnd
            }
        }
    }
`;

export const CreateSprint: React.FC = () => {
    const [sprintName, setSprintName] = React.useState<string>("");
    const [goal, setGoal] = React.useState<string>("");
    const [dateStart, setDateStart] = React.useState<string>(new Date().toISOString());
    const [dateEnd, setDateEnd] = React.useState<string>(new Date().toISOString());
    const [createSprint] = useMutation(CREATE_SPRINT);

    const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        const data = {
            sprintName,
            goal,
            dateStart: new Date(dateStart).toISOString(),
            dateEnd: new Date(dateEnd).toISOString(),
        }
        createSprint({ variables: data });

    }

    return (
        <div>
            {`Create Sprint...`}
            <form action="localhost:5557/sprint" method="post" onSubmit={handleSubmit}>
                <TextInput label="Sprint Name" name="sprintName" value={sprintName} onChange={(e: React.ChangeEvent<HTMLInputElement>) => setSprintName(e.target.value)} />
                <TextInput label="Goal" name="goal" value={goal} onChange={(e: React.ChangeEvent<HTMLInputElement>) => setGoal(e.target.value)} />
                <label htmlFor="dateInput__dateStart">Date Start</label>
                <input id="dateInput__dateStart" type="date" name="dateStart" value={dateStart.split("T")[0]} onChange={(e: React.ChangeEvent<HTMLInputElement>) => setDateStart(e.target.value)} />
                <label htmlFor="dateInput__dateEnd">Date End</label>
                <input id="dateInput__dateEnd" type="date" name="dateEnd" value={dateEnd.split("T")[0]} onChange={(e: React.ChangeEvent<HTMLInputElement>) => setDateEnd(e.target.value)} />
                <input type="submit" value="submit" />
            </form>
        </div>
    );
}