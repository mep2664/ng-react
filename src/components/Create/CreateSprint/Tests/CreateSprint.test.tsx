import React from "react"
import { render, fireEvent, wait } from "@testing-library/react"
import "@testing-library/jest-dom/extend-expect";
import { CreateSprint, CREATE_SPRINT } from "../CreateSprint"
import { MockedProvider } from '@apollo/react-testing';

const vars = { sprintName: "SPRINT_NAME", goal: "GOAL", dateStart: "2019-01-01", dateEnd: "2019-01-15" };
let numTimesSubmitted = 0;
const mocks = [
    {
        request: {
            query: CREATE_SPRINT,
            variables: { ...vars, dateStart: new Date(vars.dateStart).toISOString(), dateEnd: new Date(vars.dateEnd).toISOString() },
        },

        result: () => {
            numTimesSubmitted++;
            return {
                data: {
                    project: {
                        sprintId: "SPRINT_ID",
                        sprintName: vars.sprintName,
                        goal: vars.goal,
                        dateStart: new Date(vars.dateStart).toISOString(),
                        dateEnd: new Date(vars.dateEnd).toISOString(),
                    },
                },
            };
        },
    },
];

describe("Tests for the CreateSprint component", () => {
    test("Component matches snapshot", async () => {
        const { container } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateSprint />
            </MockedProvider>,
        );
        await wait();
        expect(container).toMatchSnapshot();
    });

    test("Changes sprint name correctly", async () => {
        const { getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateSprint />
            </MockedProvider>,
        );
        await wait();
        const input = getByLabelText("Sprint Name");
        fireEvent.change(input, { target: { value: vars.sprintName } });
        expect(input).toHaveValue(vars.sprintName);
    });

    test("Changes goal correctly", async () => {
        const { getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateSprint />
            </MockedProvider>,
        );
        await wait();
        const input = getByLabelText("Goal");
        fireEvent.change(input, { target: { value: vars.goal } });
        expect(input).toHaveValue(vars.goal);
    });

    test("Changes date start correctly", async () => {
        const { getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateSprint />
            </MockedProvider>,
        );
        await wait();
        const input = getByLabelText("Date Start");
        fireEvent.change(input, { target: { value: vars.dateStart } });
        expect(input).toHaveValue(vars.dateStart);
    });

    test("Changes date end correctly", async () => {
        const { getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateSprint />
            </MockedProvider>,
        );
        await wait();
        const input = getByLabelText("Date Start");
        fireEvent.change(input, { target: { value: vars.dateEnd } });
        expect(input).toHaveValue(vars.dateEnd);
    });

    test("Submits correctly", async () => {
        const { getByText, getByLabelText } = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <CreateSprint />
            </MockedProvider>,
        );
        await wait();

        const sprintName = getByLabelText("Sprint Name");
        fireEvent.change(sprintName, { target: { value: vars.sprintName } });

        const goal = getByLabelText("Goal");
        fireEvent.change(goal, { target: { value: vars.goal } });

        const dateStart = getByLabelText("Date Start");
        fireEvent.change(dateStart, { target: { value: vars.dateStart } });

        const dateEnd = getByLabelText("Date End");
        fireEvent.change(dateEnd, { target: { value: vars.dateEnd } });

        const submit = getByText("submit");
        fireEvent.click(submit);
        await wait();
        expect(numTimesSubmitted).toEqual(1);
    });
});
