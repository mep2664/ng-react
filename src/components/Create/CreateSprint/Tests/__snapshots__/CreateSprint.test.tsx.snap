// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the CreateSprint component Component matches snapshot 1`] = `
<div>
  <div>
    Create Sprint...
    <form
      action="localhost:5556/sprint"
      method="post"
    >
      <label
        class="sc-iAyFgw fQEosV"
        for="textInput__Sprint Name"
      >
        Sprint Name
      </label>
      <input
        class="sc-hSdWYo dqOLGd"
        id="textInput__Sprint Name"
        name="sprintName"
        type="text"
      />
      <label
        class="sc-iAyFgw fQEosV"
        for="textInput__Goal"
      >
        Goal
      </label>
      <input
        class="sc-hSdWYo dqOLGd"
        id="textInput__Goal"
        name="goal"
        type="text"
      />
      <label
        for="dateInput__dateStart"
      >
        Date Start
      </label>
      <input
        id="dateInput__dateStart"
        name="dateStart"
        type="date"
        value="2019-10-17"
      />
      <label
        for="dateInput__dateEnd"
      >
        Date End
      </label>
      <input
        id="dateInput__dateEnd"
        name="dateEnd"
        type="date"
        value="2019-10-17"
      />
      <input
        type="submit"
        value="submit"
      />
    </form>
  </div>
</div>
`;
